cpp/test/sg/rf_treelite_test.cu

```diff
@@ -0,0 +1,416 @@
+ */
+
+#include <cuda_utils.h>
+#include <gtest/gtest.h>
+#include <test_utils.h>
+#include <treelite/c_api.h>
+#include <treelite/c_api_runtime.h>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include "decisiontree/decisiontree_impl.h"
+#include "ml_utils.h"
+#include "randomforest/randomforest.hpp"
+
+/** check for system errors and assert accordingly */
+#define SYSTEMCALL_CHECK(call)                                    \
+  do {                                                            \
+    int status = call;                                            \
+    ASSERT(status == 0, "SYSTEM CALL FAIL: call='%s'.\n", #call); \
+  } while (0)
+
+namespace ML {
+
+using namespace MLCommon;
+
+template <typename T>  // template useless for now.
+struct RfInputs {
+  int n_rows;
+  int n_cols;
+  int n_trees;
+  float max_features;
+  float rows_sample;
+  int n_inference_rows;
+  int max_depth;
+  int max_leaves;
+  bool bootstrap;
+  bool bootstrap_features;
+  int n_bins;
+  int split_algo;
+  int min_rows_per_node;
+  CRITERION split_criterion;
+};
+
+template <typename T>
+::std::ostream& operator<<(::std::ostream& os, const RfInputs<T>& dims) {
+  return os;
+}
+
+template <typename T>
+class RfTreeliteTestCommon : public ::testing::TestWithParam<RfInputs<T>> {
+ protected:
+  void commonSetUp() {
+    params = ::testing::TestWithParam<RfInputs<T>>::GetParam();
+
+    DecisionTree::DecisionTreeParams tree_params;
+    set_tree_params(tree_params, params.max_depth, params.max_leaves,
+                    params.max_features, params.n_bins, params.split_algo,
+                    params.min_rows_per_node, params.bootstrap_features,
+                    params.split_criterion, false);
+    set_all_rf_params(rf_params, params.n_trees, params.bootstrap,
+                      params.rows_sample, tree_params);
+    // print(rf_params);
+
+    data_len = params.n_rows * params.n_cols;
+    inference_data_len = params.n_inference_rows * params.n_cols;
+
+    allocate(data_d, data_len);
+    allocate(inference_data_d, inference_data_len);
+
+    treelite_predicted_labels.resize(params.n_inference_rows);
+    ref_predicted_labels.resize(params.n_inference_rows);
+
+    CUDA_CHECK(cudaStreamCreate(&stream));
+    handle.setStream(stream);
+  }
+
+  void convertToTreelite(ModelHandle model) {
+    CompilerHandle compiler;
+    // "ast_navive" is the default compiler treelite used in their Python code.
+    TREELITE_CHECK(TreeliteCompilerCreate("ast_native", &compiler));
+    int verbose = 0;
+    // Generate C code in the directory specified below.
+    TREELITE_CHECK(TreeliteCompilerGenerateCode(compiler, model, verbose,
+                                                "./test_treelite"));
+    TREELITE_CHECK(TreeliteCompilerFree(compiler));
+
+    // Options copied from https://github.com/dmlc/treelite/blob/528d883f8f39eb5dd633e929b95915b63e210b39/python/treelite/contrib/__init__.py.
+    const char* obj_cmd =
```

Comment:
Can tests in gtest run in parallel? If yes, different test cases should use different filenames. Even if the tests always run sequentially, using different filenames could be useful for debugging.

Reply:
I am adding one extra parameter which indicating the index of each test case.

Reply:
I found I don't need extra parameter, `::testing::UnitTest::GetInstance()->current_test_info()->name()` can help.